### PR TITLE
fix: place J2CL inline threads at reply anchors

### LINKED ISSUE #1167: J2CL blip threading, inline blips, plus affordance, and thread-count parity
Parent: #1161\nTracker: #904\n\n## Scope\nFix blip threading and inline reply rendering against GWT.\n\n## Gaps\n- Clicking + under a blip can scroll to the bottom.\n- + affordance consumes too much vertical space and is unclear.\n- Blip threads are unclear and may not work correctly.\n- Inline blips render as replies instead of inline.\n- Diamond/thread-count widgets sit under blip and consume excessive space.\n- Collapse/expand thread semantics differ from GWT.\n\n## Acceptance\n- Inline blips, continuation/thread markers, plus affordance, and collapse/expand behavior match GWT semantics.\n- No unexpected scroll-to-bottom on plus actions.\n- Add parity harness coverage for inline blip and thread interactions.

### PR DESCRIPTION
**Pull request by vega113**
## Summary
- Render inline reply anchor placeholders inside plain-text J2CL blip content.
- Mount matching child thread hosts at those anchors for both flat and viewport-window rendering.
- Preserve the existing sibling fallback for unanchored/malformed threads and keep reply-count/collapse metadata intact.

## Verification
- RED first: `sbt --batch j2clSearchTest` failed before implementation with missing `InlineAnchorTextSegment` / `inlineAnchorTextSegmentsForTest` symbols.
- `sbt --batch j2clSearchTest` passed after implementation.
- `python3 scripts/assemble-changelog.py && python3 scripts/validate-changelog.py --fragments-dir wave/config/changelog.d --changelog wave/config/changelog.json && git diff --check && sbt --batch compile Test/compile j2clSearchTest j2clLitTest` passed after self-review tweak.

Closes #1167
Part of #904

## Summary by CodeRabbit

* **New Features**
  * Inline reply anchoring: reply threads can be mounted at placeholder anchors within blip text in both normal and windowed rendering; unmatched/malformed replies fall back to sibling placement. Existing collapse/toggle, reply-count, focus and scroll behaviors preserved.

* **Tests**
  * Added DOM and rendering tests for anchor matching, fallback placement, ordering by text offset, mention interactions, and window-render parity.

* **Documentation**
  * Added implementation plan and changelog entry.

### CODE REVIEW

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 1]
```diff
@@ -1737,25 +1848,71 @@ private HTMLElement resolveWinThreadTarget(
...
+      HTMLElement parentBlipHost,
+      HTMLElement threadHost,
+      String parentBlipId,
+      String threadId,
+      Map<String, HTMLElement> lastThreadHostByParent) {
+    HTMLElement inlineAnchor = inlineReplyAnchorHost(parentBlipHost, threadId);
+    if (inlineAnchor != null) {
+      threadHost.setAttribute("data-inline-reply-anchor-thread-id", threadId);
+      threadHost.setAttribute("data-inline-reply-anchor-placement", "inline");
+      inlineAnchor.appendChild(threadHost);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in 65a0129c7. Anchored thread hosts now mount as direct children of the parent <wave-blip> with slot="blip-extension", so they render outside the parent body slot and do not inherit task-completed body styling.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 2]
```diff
@@ -1206,6 +1242,81 @@ private void renderBlipText(
...
+
+  private static HTMLElement renderInlineReplyAnchor(InlineAnchorTextSegment segment) {
+    HTMLElement marker = (HTMLElement) DomGlobal.document.createElement("span");
+    marker.className = "j2cl-inline-reply-anchor";
+    marker.setAttribute("data-inline-reply-anchor", "true");
+    marker.setAttribute("data-inline-reply-anchor-thread-id", segment.getThreadId());
+    marker.setAttribute(
+        "data-inline-reply-anchor-offset", Integer.toString(segment.getTextOffset()));
+    marker.setAttribute("aria-label", "Inline replies");
+    return marker;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java:1333]: Fixed in 65a0129c7. The span remains only as an aria-hidden lookup marker; the thread <div> is no longer appended inside it and instead mounts on the parent <wave-blip> extension slot.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 3]
```diff
@@ -1206,6 +1242,81 @@ private void renderBlipText(
...
+  }
+
+  private static HTMLElement renderInlineReplyAnchor(InlineAnchorTextSegment segment) {
+    HTMLElement marker = (HTMLElement) DomGlobal.document.createElement("span");
+    marker.className = "j2cl-inline-reply-anchor";
+    marker.setAttribute("data-inline-reply-anchor", "true");
+    marker.setAttribute("data-inline-reply-anchor-thread-id", segment.getThreadId());
+    marker.setAttribute(
+        "data-inline-reply-anchor-offset", Integer.toString(segment.getTextOffset()));
+    marker.setAttribute("aria-label", "Inline replies");
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in 65a0129c7. Removed the aria-label and marked the empty placeholder aria-hidden so it does not enter the accessibility tree; the enhanced thread host owns the accessible group/control semantics.

File: docs/superpowers/plans/2026-05-01-issue-1167-inline-placement-parity.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+In the flat-render branch, when creating a new `threadHost`, call the helper instead of duplicating insertion logic. Keep `lastThreadHostByParent` for fallback sibling ordering only; anchored hosts are ordered by text placeholders in the parent content.
+
+- [ ] **Step 4: Update `resolveWinThreadTarget` placement**
+
+In the window-render path, use the same helper with `parentHost` and `threadId` so viewport hydration and live updates behave the same way.
+
+## Task 4: Changelog And Verification
+
+**Files:**
+- Create: `wave/config/changelog.d/20260501-j2cl-inline-placement-parity.json`
```
[the remark this answers is not in the document]
Comment (vega113) [reply docs/superpowers/plans/2026-05-01-issue-1167-inline-placement-parity.md]: Fixed in 65a0129c7. Updated the plan to reference wave/config/changelog.d/2026-05-01-j2cl-inline-placement-parity.json, matching the actual fragment naming convention.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in 872be0e9a. Added insertAnchoredThreadInOrder() which walks existing blip-extension slot children and uses insertBefore to place the new thread host before the first already-mounted thread whose anchor offset is larger. DOM order now mirrors text-offset order regardless of child-blip arrival order. Added renderPlacesMultipleAnchoredThreadsInTextOffsetOrder test that supplies two anchors in reverse arrival order and asserts the correct sequence.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 5]
```diff
@@ -1165,9 +1176,34 @@ private static void setProperty(HTMLElement element, String name, Object value)
...
+      List<J2clInlineReplyAnchor> inlineReplyAnchors) {
     String safeText = text == null ? "" : text;
+    // The current mention renderer already owns text-node splitting. Until
+    // mention and inline-reply ranges are merged into one rich-text pass, keep
+    // mention rendering authoritative and let anchored threads fall back to
+    // sibling placement for this uncommon overlap.
     if (mentions == null || mentions.isEmpty()) {
-      content.textContent = safeText;
+      List<InlineAnchorTextSegment> segments =
+          inlineAnchorTextSegments(safeText, inlineReplyAnchors);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java:1191]: Fixed in 8ff825db6. The plain `createTextNode` calls in the mention rendering path have been replaced with `appendTextSegmentWithAnchors()`, which iterates sorted anchors within each text gap and emits `<span data-inline-reply-anchor>` markers interleaved with text nodes. Blips with both mention chips and inline anchors now produce both marker spans and correctly mounted thread hosts. Test `renderBlipTextRendersInlineAnchorSpansEvenWhenMentionsArePresent` added.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 6]
```diff
@@ -1183,7 +1216,7 @@ private void renderBlipText(
         continue;
       }
       if (start > cursor) {
-        content.appendChild(DomGlobal.document.createTextNode(safeText.substring(cursor, start)));
+        appendTextSegmentWithAnchors(content, safeText, cursor, start, sortedAnchors);
       }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in 16466e6fd. Removed the `if (start > cursor)` guard before `appendTextSegmentWithAnchors` so the helper is always invoked (including when `start == cursor`). Combined with the boundary fix below, anchors exactly at a mention's `startOffset` are now emitted as marker spans. Test `renderBlipTextRendersInlineAnchorAtMentionStartBoundary` added.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 7]
```diff
@@ -1197,15 +1230,113 @@ private void renderBlipText(
       renderedMention = true;
     }
     if (cursor < safeText.length()) {
-      content.appendChild(DomGlobal.document.createTextNode(safeText.substring(cursor)));
+      appendTextSegmentWithAnchors(content, safeText, cursor, safeText.length(), sortedAnchors);
     }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in 16466e6fd. Removed the `if (cursor < safeText.length())` guard on the trailing call, so `appendTextSegmentWithAnchors` is always invoked after the mention loop. The boundary check in the helper was also widened from `offset >= to` to `offset > to` so anchors at exactly `safeText.length()` are processed instead of skipped. Test `renderBlipTextRendersInlineAnchorAtEndOfTextWithMentions` added.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 8]
```diff
@@ -1197,15 +1230,113 @@ private void renderBlipText(
...
+      HTMLElement content,
+      String safeText,
+      int from,
+      int to,
+      List<J2clInlineReplyAnchor> sortedAnchors) {
+    int cursor = from;
+    for (J2clInlineReplyAnchor anchor : sortedAnchors) {
+      int offset = normalizedAnchorOffset(safeText, anchor);
+      if (offset < from || offset >= to) {
+        continue;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java:1246]: Fixed in 16466e6fd. Changed `offset >= to` to `offset > to` as suggested, making the upper boundary inclusive so anchors at exactly `to` are processed. Tests added for both boundary cases.
